coderd/database/pubsub/pubsub.go

```diff
@@ -239,8 +239,11 @@ func (p *pgPubsub) Publish(event string, message []byte) error {
 
 // Close closes the pubsub instance.
 func (p *pgPubsub) Close() error {
-	p.cancel()
+	// canceling the context also causes the listener to be closed, so directly
+	// close it before canceling the context to avoid the race. If we lost the
+	// race we'd return an error about the listener already being closed.
 	err := p.pgListener.Close()
+	p.cancel()
```

Comment:
In theory, this way we're introducing a race in `Publish` where the listener might be closed whilst we succeed in publishing due to context still being active. I've not tried to deduce whether or not this is significant, just noting the behavior.

Is there a reason we don't simply test for "already closed" here instead? Or unify the close paths into a `sync.Once` or similar?

Reply:
canceling a context is always async to stopping other goroutines that check it, so I don't think we've fundamentally changed any race between `Close()` and `Publish()`.  

Testing for already closed is annoying because `pq` doesn't export the error, so we'd have to string match.  I guess synchronizing the close is simple enough.  I don't want to use a `Once` though because it'll make it fiddly to get an error, just a mutex and `closed bool`...